File: Classes/GameEngine/EngineState.m
//
//  EngineState.m
//  AirHockey
//
//  Created by Jonathan Sharkey on 10-04-10.
//  Copyright 2010 Sharkable. All rights reserved.
//

#import "EngineState.h"
#import "GameEngine.h"

@implementation EngineState

- (id) init {
	[super init];
	
	_entities = [[NSMutableArray alloc] init];
	
	return self;
}

- (void) dealloc {
	[_entities release];
	
	[super dealloc];
}

- (void) stateIsShown {
}

- (void) update {
	for (id <StateEntity> entity in _entities) {
		[entity update];
	}
}

- (void) render {
	for (id <StateEntity> entity in _entities) {
		[entity render];
	}
}

- (void) addEntity:(id <StateEntity>)entity {
	[_entities addObject:entity];
}

- (void) removeEntity:(id <StateEntity>)entity {
	[_entities removeObject:entity];
}

- (void) touchesBegan:(Touch*[])touches numTouches:(int)numTouches {
	for (NSObject <StateEntity>* entity in _entities) {
		if ([entity respondsToSelector:@selector(touchesBegan:numTouches:)]) {
			[entity touchesBegan:touches numTouches:numTouches];
		}
	}
}

- (void) touchesMoved:(Touch*[])touches numTouches:(int)numTouches {
	for (NSObject <StateEntity>* entity in _entities) {
		if ([entity respondsToSelector:@selector(touchesMoved:numTouches:)]) {
			[entity touchesMoved:touches numTouches:numTouches];
		}
	}
}

- (void) touchesEnded:(Touch*[])touches numTouches:(int)numTouches {
	for (NSObject <StateEntity>* entity in _entities) {
		if ([entity respondsToSelector:@selector(touchesEnded:numTouches:)]) {
			[entity touchesEnded:touches numTouches:numTouches];
		}
	}
}

- (void) clearTouches {
	for (NSObject <StateEntity>* entity in _entities) {
		if ([entity respondsToSelector:@selector(clearTouches)]) {
			[entity clearTouches];
		}
	}
}

@end

File: Classes/GameEngine/EngineState.cc
//
//  EngineState.m
//  AirHockey
//
//  Created by Jonathan Sharkey on 10-04-10.
//  Copyright 2010 Sharkable. All rights reserved.
//

#import "EngineState.h"
#import "GameEngine.h"

EngineState::EngineState(GameEngine *gameEngine) {
  gameEngine_ = gameEngine;
}

void EngineState::stateIsShown() {
}

void EngineState::update() {
  for (int i = 0; i < entities_.size(); i++) {
    entities_[i]->update();
  }
}

void EngineState::render() {
  for (int i = 0; i < entities_.size(); i++) {
    entities_[i]->render();
  }
}

void EngineState::addEntity(StateEntity *entity) {
  entities_.push_back(entity);
}

void EngineState::removeEntity(StateEntity *entity) {
  for (vector<StateEntity *>::iterator i = entities_.begin();
      i != entities_.end(); i++) {
    if (*i == entity) {
      entities_.erase(i);
      break;
    }
  }
}

void EngineState::touchesBegan(Touch *touches[], int numTouches) {
  for (int i = 0; i < entities_.size(); i++) {
    entities_[i]->touchesBegan(touches, numTouches);
  }
}

void EngineState::touchesMoved(Touch *touches[], int numTouches) {
  for (int i = 0; i < entities_.size(); i++) {
    entities_[i]->touchesMoved(touches, numTouches);
  }
}

void EngineState::touchesEnded(Touch *touches[], int numTouches) {
  for (int i = 0; i < entities_.size(); i++) {
    entities_[i]->touchesEnded(touches, numTouches);
  }
}

void EngineState::clearTouches() {
  for (int i = 0; i < entities_.size(); i++) {
    entities_[i]->clearTouches();
  }
}
